doc(cn) - translate form (#95)
reactjs/zh-hans.reactjs.org: content/docs/forms.md

  - "docs/forms-zh-CN.html"
---

> 

Reviewer: 标题译为表单

</form>
```

> 此表单具有默认的 HTML 表单行为，即用户提交表单后浏览到新页面。在 React 中，如果你只需要这种行为，你不用做任何修改。但是，在大多数情况下，拥有一个处理表单提交和对用户输入的数据的访问的 javascript 函数是很方便的。实现这一点的标准方法是使用一种称为“受控组件”的技术。

Reviewer 2: 此表单具有默认的 HTML 表单行为，即在用户提交表单后浏览到新页面。如果你想在 React 中实现这种行为，你无需做任何修改。但大多数情况下，使用 JavaScript 函数来处理表单提交、访问用户填写的表单数据会更方便。而比较标准的实现方法是使用一种称为“受控组件”的技术。

- 👍点赞「你无需做任何修改」翻译；
- JavaScript 大小写要规范哈；
- 长句尽量改为更易理解的多个短句；

## 受控组件 {#controlled-components}

> 在HTML中，表单元素（如`<input>`、`<textarea>`和`<select>`）通常保持自己的 state，并根据用户输入进行更新。在 React 中，可变状态通常保存在组件的 state 属性中，并且只用 [`setState()`](/docs/react-component.html#setstate)来更新.

Reviewer 2: 在 HTML 中，表单元素（如 `<input>`、`<textarea>` 和 `<select>`）通常会维护自己的 state, 并根据用户输入进行更新。而在 React 中，可变状态通常保存在组件的 state 属性中，并且只通过 [`setState()`](/docs/react-component.html#setstate) 进行更新。

- 中英文之间加空格哈；
- maintain 这里感觉翻译成「维护」更好哈；
- 一般我们说「通过 setState() 更新 state」；

> 我们可以把两者结合起来，使 React state 成为“事实的维一来源”。这样控制表单渲染的 React 组件还控制在随后的用户输入对表单的影响。这种输入值由 React 以这种方式控制着表单元素称为“受控组件”。

Reviewer 2: 我们可以通过把 React state 作为“事实的唯一来源”，来将二者合并起来。这样渲染表单的 React 组件同时也控制了用户后续输入对表单造成的影响。被 React 以这种方式控制取值的表单输入元素就叫做“受控组件”。

> 例如，如果我们想让前一个示例在提交时记录名称，我们可以将表单作为受控组件写入：

Reviewer 2: 例如，如果我们想让前一个示例在提交时打印出名称，我们可以将表单写为受控组件：

- 从下文 handleSubmit 方法可看出是把名称 alert 了出来，所以不妨翻译为「打印」；

}
```

[**在CodePen试一试**](https://codepen.io/gaearon/pen/VmmPgp?editors=0010)

由于在表单元素上设置了 `value` 属性，因此显示的值将始终为 `this.state.value`，这使React state成为事实的来源。由于 `handlechange` 在每次按键时都会运行以更新 React state，因此显示的值将随着用户输入而更新。

对于受控组件，每个状态突变都有一个相关的处理函数。这使得修改或验证用户输入变得简单。例如，如果我们要强制要求所有名称都用大写字母书写，我们可以将 `handlechange` 写为：

```javascript{2}
handleChange(event) {
  this.setState({value: event.target.value.toUpperCase()});
}
```

## textarea 标签 {#the-textarea-tag}

在HTML中, `<textarea>` 元素通过其子元素定义其文本:

```html
<textarea>
  Hello there, this is some text in a text area
</textarea>
```

> 在 React 中，`<textarea>` 用 `value` 属性。这样，使用 `<textarea>` 的表单可以和使用单行input的表单非常类似：

Reviewer 2: 而在 React 中，`<textarea>` 使用的是 `value` 属性。这样，使用 `<textarea>` 的表单可以和使用单行 input 的表单非常相似：

- 可以适当使用连词及补充，使上下文更加连贯；

</select>
```

> 请注意，由于 `selected` 属性的缘故，Coconut 选项选项默认被选中。React 并不会使用 `selected` 属性，而是在根 `select` 标签上使用 `value` 属性。这在受控组件中更便捷，因为您只需要在根标签中更新它。例如：

Reviewer: 多了一个选项

> </select>

Reviewer 3: 代码中内容是否要统一翻译？我记得之前有 PR 针对其中内容是进行翻译的。

Reviewer: @zhangxd6 代码块中的英文也翻译一下，辛苦

}
```

[**在CodePen试一试**](https://codepen.io/gaearon/pen/JbbEzX?editors=0010)

总的来说，这使得 `<input type=“text”>`, `<textarea>` 和 `<select>` 所有的标签工作都非常相似-它们都接受一个 `value` 属性，您可以使用它来实现一个受控制的组件。

> 注意
>

> > 可以将数组传递到 `value` 属性中，允许您在 `select` 标签中选择多个选项：

Reviewer 2: 你可以将数组传递到 `value` 属性中，以支持在 `select` 标签中选择多个选项：

- 尽量贴近中文表达习惯吧；

>
>```js
><select multiple={true} value={['B', 'C']}>
>```

## 文件 input 标签 {#the-file-input-tag}

> 在HTML中，`<input type=“file”>` 允许用户从设备存储中选择一个或多个文件上载到服务器或通过 [File API](https://developer.mozilla.org/en-US/docs/Web/API/File/Using_files_from_web_applications) 来控制。

Reviewer 2: 在HTML中，`<input type=“file”>` 允许用户从设备存储中选择一个或多个文件，将其上传到服务器，或使用 JavaScript 的 [File API](https://developer.mozilla.org/en-US/docs/Web/API/File/Using_files_from_web_applications) 进行控制。

}
```

[**在CodePen试一试**](https://codepen.io/gaearon/pen/wgedvV?editors=0010)

> 请注意，我们是如何使用 ES6 [计算属性名称](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Operators/Object_initializer#Computed_property_names)语法更新与给定输入名称对应的 state 键：

Reviewer 2: 请注意，我们是如何使用 ES6 的 [计算属性名称](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Operators/Object_initializer#Computed_property_names)语法使 state 的键名随着给定的输入名更新：

> 请注意，我们是如何使用 ES6 [计算属性名称](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Operators/Object_initializer#Computed_property_names)语法更新与给定输入名称对应的 state 键：

Reviewer: ```suggestion
这里使用了 ES6 [计算属性名称](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Operators/Object_initializer#Computed_property_names) 的语法更新给定输入名称对应的 state 值：
```

例如：

```js{2}
this.setState({
  [name]: value
});
```

等同 ES5:

```js{2}
var partialState = {};
partialState[name] = value;
this.setState(partialState);
```

> 另外，由于 `setState()` 自动[将部分状态合并到当前状态](/docs/state-and-lifecycle.html#state-updates-are-merged), 我们只需要用更改的部分来调用它。

Reviewer: ```suggestion
另外，由于 `setState()` 自动[将部分 state 合并到当前 state](/docs/state-and-lifecycle.html#state-updates-are-merged)，只需调用它更改部分 state 即可。
```

## 受控输入空值 {#controlled-input-null-value}

> 如果您愿意，在[受控元素](/docs/forms.html#controlled-components)上指定 value prop 可防止用户更改输入。如果指定了 `value`，但输入仍可编辑，则可能是意外地将`value` 设置为  `undefined`或 `null`。

Reviewer 2: 在[受控元素](/docs/forms.html#controlled-components)上指定 value 属性可防止用户更改 input, 除非你想这么做。如果你已经指定了 `value`，但 input 仍可编辑，那你可能不小心将 `value` 设置为 `undefined` 或 `null` 了。

> 如果您愿意，在[受控元素](/docs/forms.html#controlled-components)上指定 value prop 可防止用户更改输入。如果指定了 `value`，但输入仍可编辑，则可能是意外地将`value` 设置为  `undefined`或 `null`。

Reviewer: ```suggestion
在[受控组件](/docs/forms.html#controlled-components)上指定 value 的 prop 可防止用户更改输入。如果指定了 `value`，但输入仍可编辑，则可能会意外地将`value` 设置为 `undefined` 或 `null`。
```

[**在CodePen试一试**](https://codepen.io/gaearon/pen/wgedvV?editors=0010)

> 请注意，我们是如何使用ES6[计算属性名称](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Operators/Object_initializer#Computed_property_names)语法更新与给定输入名称对应的state键：

Reviewer 4: “语法更新与给定输入名称对应的state键”

中英文之间应该有空格（React，state等词）

```

## 受控组件的替代品 {#alternatives-to-controlled-components}

> 有时使用受控制的组件会很麻烦，因为您需要为数据可以更改的每种方式编写一个事件处理程序，并通过一个 React 组件传递所有的输入状态。当您将先前存在的代码库转换为 React 或将 React 应用程序与非 React 库集成时，这可能会变得特别烦人。在这些情况下，您可能希望使用[非受控组件](/docs/uncontrolled-components.html), 这是实现输入表单的另一种技术。

Reviewer: ```suggestion
有时使用受控组件会很麻烦，因为你需要为数据变化的每种方式都编写事件处理函数，并通过 React 组件传递所有的输入 state。当你将之前的代码库转换为 React 或将 React 应用程序与非 React 库集成时，这可能会令人厌烦。在这些情况下，你可能希望使用[非受控组件](/docs/uncontrolled-components.html), 这是实现输入表单的另一种方式。
```

## 成熟的解决方案 {#fully-fledged-solutions}

> 如果您正在寻找一个完整的解决方案，包括验证、跟踪访问的字段和处理表单提交, [Formik](https://jaredpalmer.com/formik) 是一个流行的选择。然而，它建立在受控组件和管理状态的相同原则之上——所以不要忽视学习它们。

Reviewer: ```suggestion
如果你想寻找包含验证、追踪访问字段以及处理表单提交的完整解决方案，使用 [Formik](https://jaredpalmer.com/formik) 是不错的选择。然而，它也是建立在受控组件和管理 state 的基础之上——所以不要忽视学习它们。
```

```

## 受控组件的替代品 {#alternatives-to-controlled-components}

有时使用受控组件会很麻烦，因为你需要为数据变化的每种方式都编写事件处理函数，并通过一个 React 组件传递所有的输入 state。当你将之前的代码库转换为 React 或将 React 应用程序与非 React 库集成时，这可能会令人厌烦。在这些情况下，你可能希望使用[非受控组件](/docs/uncontrolled-components.html), 这是实现输入表单的另一种方式。

## 成熟的解决方案 {#fully-fledged-solutions}

> 如果你想寻找包含验证、追踪访问字段以及处理表单提交的完整解决方案，使用 [Formik](https://jaredpalmer.com/formik) 是不错的选择。然而，它也是建立在受控组件和管理 state 的基础之上——所以不要忽视学习它们。

Reviewer 2: 如果你正在寻找一个包含验证、已访问字段追踪以及表单提交处理的完整解决方案，[Formik](https://jaredpalmer.com/formik) 是个不错的选择。不过，它也是基于受控组件和 state 管理的原理之上构建的——所以不要忽视对它们的学习。

- 英文中很多动词的表达，翻译成名词会更符合我们的表达习惯；
